File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        cur_sum = 0
        prefix_sum = { 0: 1}
        res = 0
        
        for n in nums:
            cur_sum += n
            diff = cur_sum - k
            
            res += prefix_sum.get(diff, 0)
            prefix_sum[cur_sum] = 1 + prefix_sum.get(cur_sum, 0)
        return res

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    let prefixSum = 0;
    let res = 0;
    let prefixMap = new Map();
    prefixMap.set(0, 1);
    
    // prefixSum - sum = k
    
    for (let num of nums) {
        prefixSum += num;
        const potentialAns = prefixSum - k;
        if (prefixMap.has(potentialAns)) {
            res += prefixMap.get(potentialAns);
        }
        prefixMap.set(prefixSum, (prefixMap.get(prefixSum) || 0) + 1)
    }
    
    return res
    

    
};